projects/pgai/pgai/semantic_catalog/gen_sql.py

```diff
@@ -554,127 +719,166 @@ async def generate_sql(
+                            list[int] | None, args.get("relevant_fact_ids")
+                        )
+                        if relevant_object_ids:
+                            for irrelevant_id in (
+                                ctx.objects.keys() - relevant_object_ids
+                            ):
+                                ctx.objects.pop(irrelevant_id)
+                                ctx.rendered_objects.pop(irrelevant_id)
+                            await _increment_obj_usage(
+                                catalog_con,
+                                catalog_id,
+                                [id for id in ctx.objects],
+                            )
+                        if relevant_sql_example_ids:
+                            for irrelevant_id in (
+                                ctx.sql_examples.keys() - relevant_sql_example_ids
+                            ):
+                                ctx.sql_examples.pop(irrelevant_id)
+                                ctx.rendered_sql_examples.pop(irrelevant_id)
+                            await _increment_sql_usage(
+                                catalog_con,
+                                catalog_id,
+                                [id for id in ctx.sql_examples],
+                            )
+                        if relevant_fact_ids:
+                            for irrelevant_id in ctx.facts.keys() - relevant_fact_ids:
+                                ctx.facts.pop(irrelevant_id)
+                                ctx.rendered_facts.pop(irrelevant_id)
+                            await _increment_fact_usage(
+                                catalog_con, catalog_id, [id for id in ctx.facts]
+                            )
+                        command_type = str(args["command_type"])
+                        if command_type.upper() in (
+                            "SELECT",
+                            "INSERT",
+                            "UPDATE",
+                            "DELETE",
+                            "MERGE",
+                            "VALUES",
+                        ):
+                            logger.info("validating the sql statement")
+                            query_plan, error = await validate_sql_statement(
+                                target_con, str(answer)
+                            )
+                            if error:
+                                logger.info(
+                                    "asking llm to fix the invalid sql statement"
+                                )
+                                _continue = True
+                    else:
+                        logger.error(
+                            f"unrecognized tool name: {tool_call_part.tool_name}"
+                        )
+                case "text":
+                    text_part: TextPart = part
+                    logger.info(f"model response: {text_part.content}")
```

Comment:
As we're specifying `allow_text_output=False,` to the request options, hitting this case _should_ be impossible, right?

Reply:
eb3d0daa82733ae52115567b2e8fde045f89c1dd